oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentStoreThrottlingMetricsUpdater.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.document.mongo;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.math.DoubleMath;
+import com.google.common.util.concurrent.AtomicDouble;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import org.bson.BsonTimestamp;
+import org.bson.Document;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Mongo Document Store throttling metric updater.
+ *
+ * This class fetches and updates the mongo oplog window
+ */
+public class MongoDocumentStoreThrottlingMetricsUpdater {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoDocumentStoreThrottlingMetricsUpdater.class);
+    static final String TS_TIME = "ts";
+    private static final String NATURAL = "$natural";
+    private static final String MAX_SIZE = "maxSize";
+    private static final String OPLOG_RS = "oplog.rs";
+    public static final String SIZE = "size";
+    private final ScheduledExecutorService throttlingMetricsExecutor;
+    private final AtomicDouble oplogWindow;
+    private final MongoDatabase localDb;
+
+    public MongoDocumentStoreThrottlingMetricsUpdater(final @NotNull MongoDatabase localDb, final @NotNull AtomicDouble oplogWindow) {
+        // exiting scheduled executor, will exit when we call to shut down jvm
+        this.throttlingMetricsExecutor = MoreExecutors.getExitingScheduledExecutorService(
+                (ScheduledThreadPoolExecutor) Executors.newScheduledThreadPool(1), 40, SECONDS);
+        this.oplogWindow = oplogWindow;
+        this.localDb = localDb;
+    }
+
+    public void updateMetrics() {
```

Comment:
Was wondering about the 3 warn cases : while it's unclear why those problems exist, it might be worth setting the `oplogWindow` to max_integer in that case - as basically we don't know how long such error situations last and what the impact is. So the fallback would be to say : "unknown oplog window", thus max_integer
(currently it leaves the `oplogWindow` unchanged in warn cases, which might mean throttling even though perhaps in the meantime the oplog window grew - of course the opposite is also true, but given this is an exceptional situation I'd steer to the infinite side)